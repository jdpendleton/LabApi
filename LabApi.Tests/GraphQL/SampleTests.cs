using HotChocolate.Execution;
using LabApi.Business.Interfaces;
using LabApi.Business.Services;
using LabApi.Data.Interfaces;
using LabApi.Data.Repositories;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;

namespace LabApi.Tests.GraphQL
{
    public class SampleTests
    {
        [Fact]
        public async Task SchemaChangeTest()
        {
            #region Arrange



            #endregion

            #region Act

            var schema = await new ServiceCollection()
                .AddGraphQLServer()
                .AddQueryType(d => d.Name("Query"))
                    .AddType<SampleQueries>()
                .AddMutationType(d => d.Name("Mutation"))
                    .AddType<SampleMutations>()
                .BuildSchemaAsync();

            #endregion

            #region Assert

            schema.ToString().MatchSnapshot();

            #endregion
        }

        [Fact]
        public async Task CreateSample_Snapshot()
        {
            #region Arrange

            var query = $@"
                mutation {{
                    createSample (
                        request: {{
                            name: ""a4dac2d0-eda7-4a47-aabc-809b201364aa""
                            owner: ""a4dac2d0-eda7-4a47-aabc-809b201364aa""
                            position: 1 }}
                        ) {{
                        sample {{
                            name
                        }}
                    }}
                }}";

            #endregion

            #region Act

            var result = await new ServiceCollection()
                .AddSingleton<ISampleRepository, SampleRepository>()
                .AddSingleton<ISampleService, SampleService>()
                .AddGraphQLServer()
                .AddQueryType(d => d.Name("Query"))
                    .AddType<SampleQueries>()
                .AddMutationType(d => d.Name("Mutation"))
                    .AddType<SampleMutations>()
                .AddErrorFilter<SampleNotFoundExceptionFilter>()
                .AddErrorFilter<DuplicateSampleExceptionFilter>()
                .AddErrorFilter<SampleCheckedOutExceptionFilter>()
                .AddErrorFilter<SampleArchivedExceptionFilter>()
                .AddErrorFilter<RedundantSampleRequestExceptionFilter>()
                .ExecuteRequestAsync(query);

            #endregion

            #region Assert

            result.ToJson().MatchSnapshot();

            #endregion
        }
    }
}