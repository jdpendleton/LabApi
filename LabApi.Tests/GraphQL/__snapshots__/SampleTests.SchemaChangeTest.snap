schema {
  query: Query
  mutation: Mutation
}

type ArchiveSampleResponse {
  sample: SampleDTO
  error: String
}

type CheckinSampleResponse {
  sample: SampleDTO
  error: String
}

type CheckoutSampleResponse {
  sample: SampleDTO
  error: String
}

type CreateSampleResponse {
  sample: SampleDTO
  error: String
}

type MoveSampleResponse {
  sample: SampleDTO
  error: String
}

type Mutation {
  createSample(request: CreateSampleRequestInput!): CreateSampleResponse!
  checkoutSample(request: CheckoutSampleRequestInput!): CheckoutSampleResponse!
  checkinSample(request: CheckinSampleRequestInput!): CheckinSampleResponse!
  moveSample(request: MoveSampleRequestInput!): MoveSampleResponse!
  archiveSample(request: ArchiveSampleRequestInput!): ArchiveSampleResponse!
}

type Query {
  samples: [SampleDTO!]!
}

type SampleDTO {
  name: String!
  owner: UUID!
  checkedOut: Boolean!
  position: Int!
  archived: Boolean!
}

input ArchiveSampleRequestInput {
  name: String!
}

input CheckinSampleRequestInput {
  name: String!
}

input CheckoutSampleRequestInput {
  name: String!
}

input CreateSampleRequestInput {
  name: String!
  owner: UUID!
  checkedOut: Boolean
  position: Int!
  archived: Boolean
}

input MoveSampleRequestInput {
  name: String!
  position: Int!
}

scalar UUID @specifiedBy(url: "https:\/\/tools.ietf.org\/html\/rfc4122")
